### PR TITLE
feat: retrieve additional data

### LINKED ISSUE #26: add user/appInstance data to written file
currently, the file created when a data request is triggered writes an array of action objects, plus metadata that includes a `spaceTree` (also an array of objects). we want to add to this file all users in a space and its subspaces, and the appInstances in those spaces.

### PR DESCRIPTION
**Pull request by hagoptaminian**
both api call and data file now include 'users' and 'apps' data

closes #26

### CODE REVIEW

File: controllers/tasks.js [hunk 1]
```diff
@@ -126,21 +126,36 @@ const createTask = [
...
+    // each 'memberships' key holds an array of objects, each of the format { userId: <mongoId> }
+    // i.e. usersArray = [ { memberships: [ { userId: <mongoId> }, ... ] }, ... ]
+    let userIds = [];
+    usersArray.forEach(({ memberships }) => {
+      if (memberships) {
+        memberships.forEach(({ userId }) => {
+          userIds.push(userId.toString());
+        });
+      }
+    });
```
Comment (ahgentil) [inline controllers/tasks.js:143]: another way of doing this:

```javascript
let userIds = usersArray.reduce((acc, { memberships = [] }) =>
  acc.concat(memberships.map(({ userId }) => userId.toString())),
[]);
```

**Review by ahgentil — APPROVED** [review]
just added a suggestion, for you to compare.
other than that, lgtm (**l**ooks **g**ood **t**o **m**e - if you didn't know already).

File: controllers/tasks.js (hunk 1 above)
Comment (hagoptaminian) [reply controllers/tasks.js:143]: I tested this function versus the two `.forEach()` loops and it is **significantly** slower. E.g. for a `usersArray` of length 50000, this function takes ~24 seconds to complete, while for the same `usersArray` the two `.forEach()` loops complete in <100ms. I also find the `.forEach()` syntax easier to read. So I will keep as is.

Comment (ahgentil) [reply controllers/tasks.js:143]: The slowness I didn't know. You can use the online took jsperf to test those. It's probably even faster with simple for()'s
The readability, I agree.

**Comment by hagoptaminian** [comment]
> just added a suggestion, for you to compare.
> other than that, lgtm (**l**ooks **g**ood **t**o **m**e - if you didn't know already).

@ahgentil Thanks for this detailed review. I spoke to @juancarlosfarah and I will merge this PR into master. If you have some time, please release these changes in Graasp API. Otherwise @juancarlosfarah can do it tonight/tomorrow.

**Pull request merged**